04-cs-fundamentals/classroom/Finite-State-Machines.md

```diff
@@ -0,0 +1,182 @@
+  States
+
+  Current State | State symbols
+  --- | --- 
+  Farm lights Red - Highway Green | FR - HG
+  Farm lights Yellow - Highway Red | FY - HR
+  Farm lights Green - Highway Red | FG - HR
+  Farm lights Red - Highway Yellow | FR - HY
+
+  Inputs (Alphabet)
+
+  Input | State symbols
+  --- | --- 
+  Farm Sensor - Timer Expired | FS - TEX
+  Farm Sensor - Timer Running | FS - TR
+  No Farm Sensor - Timer Running | NFS - TR
+  No Farm Sensor - Timer Expired | NFS - TEX
+
+**Bonus**
+
+Can you write a FSM which will determine if given the alphabet {a, b} that the input has an even number of the letter `a` or `b`?
+
+## Summary
+
+A Finite State Machine is a mathematical way to model computation.  The model consists of a limited number of states and a finite alphabet of possible input characters.  The machine transitions from state to state as it reads in each item of input.  An FSM does not remember previous input, but only the current state it is in.  
+
+FSMs are used in a variety of applications and can be an effective way to model and understand systems which consist of a limited number of potential states.  
+
+## Further Concepts
+
+An FSM isn't a complete computational system.  Full computers can **remember** more than one particular value.  To model this mathematicains have developed additional systems which build on FSMs.
+
+### Nondeterministic Finite State Machines
+
+A nondeterministic finite state machine is a machine which can take multiple paths given a specific input.  The machine will accept input if there is *at least* one path it can follow to accept the input.  The FSMs we've been using are called Deterministic Finite State Machines as they can only take one path for each given input.  It's been shown that you can build a deterministic finite state machine that is equivalent to a nondeterministic state machine.
+
+### Pushdown Automata
+
+A Pushdown Automata is a finite state machine that also has access to a **stack**.  Because it can access the stack to remember past inputs or states a Pushdown Automata is more powerful than a finite state machine.
```

Comment:
Give an example of something a PDA can accept that a DFA can't?  Maybe the language of balanced parenthesis?